src/metabase/util/analytics.clj

```diff
@@ -0,0 +1,119 @@
+                     (.setConnectionManager (PoolingHttpClientConnectionManager.))
+                     (.build))
+          builder (-> (ApacheHttpClientAdapter/builder)
+                      (.httpClient client)
+                      (.url (snowplow-url)))
+          adapter (.build ^ApacheHttpClientAdapter$Builder builder)
+          batch-emitter-builder (-> (BatchEmitter/builder)
+                                    (.bufferSize 1)
+                                    (.httpClientAdapter adapter))]
+      (.build ^BatchEmitter$Builder batch-emitter-builder))))
+
+(def ^:private ^Tracker tracker
+  "An instance of a Snowplow tracker"
+  (delay
+   (-> (Tracker$TrackerBuilder. ^Emitter @emitter "sp" "metabase")
+       (.build))))
+
+(defn- subject
+  "Create a Subject object for a given user ID, to be included in analytics events"
+  [user-id]
+  (-> (Subject$SubjectBuilder.)
+      (.userId (str user-id))
+      (.build)))
+
+(defn- context
+  "Common context included in every analytics event"
+  []
+  (new SelfDescribingJson
+       "iglu:com.metabase/instance/jsonschema/1-0-0"
+       {"id"             (public-settings/analytics-uuid),
+        "version"        {"tag" (:tag (public-settings/version))},
+        "token-features" (into {} (for [[token enabled?] (public-settings/token-features)]
+                                    [(name token) enabled?]))}))
+
+(defn- payload
+  "A SelfDescribingJson object containing the provided event data, which can be included as the payload for an
+  analytics event"
+  [schema version event-data]
+  (new SelfDescribingJson
+       (format "iglu:com.metabase/%s/jsonschema/%s" (name schema) version)
+       ;; Make sure keywords are converted to strings
+       (into {} (for [[k v] event-data] [(name k) (if (keyword? v) (name v) v)]))))
+
+(defn- track-schema-event
+  "Send a single analytics event to the Snowplow collector, if tracking is enabled for this MB instance"
+  [schema version user-id event-data]
+  (when (public-settings/anon-tracking-enabled)
+    (try
+      (let [^Unstructured$Builder builder (-> (. Unstructured builder)
+                                              (.eventData (payload schema version event-data))
+                                              (.customContext [(context)]))
+            ^Unstructured$Builder builder' (if user-id
+                                             (.subject builder (subject user-id))
+                                             builder)
+            ^Unstructured event (.build builder')]
+        (.track ^Tracker @tracker event))
+      (catch Throwable e
+        (log/debug e (trs "Error sending Snowplow analytics event {0}" (name (:event event-data))))))))
+
+;; Snowplow analytics interface
+
+(defmulti track-event
+  "Send a single analytics event to Snowplow"
+  (fn [event & _] event))
+
+(defmethod track-event :new_instance_created
+  [event]
+  (track-schema-event :account "1-0-0" nil {:event event}))
```

Comment:
When are we planning on passing something other than `1-0-0`? It seems like a bit of a magic number and it's repeated for every call to `track-schema-event`, does it make sense to make this an optional param?

Reply:
If we update any of the schemas in SnowcatCloud we'll need to update the version numbers when we start sending data that conforms to the new schemas. I've consolidated these into a single map and added a comment to clarify.

(Current schema definitions are checked into the https://github.com/metabase/iglu-schema-registry repo, but being moved to the main repo in #19040)